File: src/main/java/net/alexwells/roomery/gui/GuiHandler.java
package net.alexwells.roomery.gui;

import net.alexwells.roomery.mechanic.roomholder.RoomHolderGui;
import net.alexwells.roomery.mechanic.roomholder.RoomHolderContainer;
import net.alexwells.roomery.util.WorldUtils;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.network.IGuiHandler;

import javax.annotation.Nullable;

public class GuiHandler implements IGuiHandler {
    @Nullable
    @Override
    public Object getServerGuiElement(int id, EntityPlayer player, World world, int x, int y, int z) {
        if(id == GuiEnum.ROOM_HOLDER.id) {
            return new RoomHolderContainer(player.inventory, WorldUtils.getTileEntity(world, new BlockPos(x, y, z)));
        }

        return null;
    }

    @Nullable
    @Override
    public Object getClientGuiElement(int id, EntityPlayer player, World world, int x, int y, int z) {
        if(id == GuiEnum.ROOM_HOLDER.id) {
            return new RoomHolderGui(player.inventory, WorldUtils.getTileEntity(world, new BlockPos(x, y, z)));
        }

        return null;
    }
}

File: src/main/java/net/alexwells/roomery/gui/GuiHandler.kt
package net.alexwells.roomery.gui

import net.alexwells.roomery.mechanic.roomholder.ROOM_HOLDER_RESOURCE
import net.alexwells.roomery.mechanic.roomholder.RoomHolderContainer
import net.alexwells.roomery.mechanic.roomholder.RoomHolderGui
import net.alexwells.roomery.mechanic.roomholder.RoomHolderTileEntity
import net.alexwells.roomery.util.getTile
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiScreen
import net.minecraft.inventory.Container
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.world.IInteractionObject
import net.minecraftforge.fml.network.FMLPlayMessages

object GuiHandler {
    fun handleGui(packet: FMLPlayMessages.OpenContainer): GuiScreen? {
        return when(packet.id) {
            ROOM_HOLDER_RESOURCE -> {
                val container = buildContainerForTile<RoomHolderTileEntity, RoomHolderContainer>(packet.additionalData.readBlockPos()) ?: return null

                return RoomHolderGui(container)
            }
            else -> null
        }
    }

    private fun <T, C: Container> buildContainerForTile(pos: BlockPos): C? where T: TileEntity, T: IInteractionObject {
        val tile = Minecraft.getInstance().world.getTile<T>(pos) ?: return null

        val player = Minecraft.getInstance().player

        return tile.createContainer(player.inventory, player) as C
    }
}